core/src/Platforms/iOS/SystemWebview/SystemWebUI.cs

```diff
@@ -73,23 +74,77 @@ public void Authenticate(Uri authorizationUri, Uri redirectUri, UIViewController
+                if (UIDevice.CurrentDevice.CheckSystemVersion(12, 0))
+                {
+                    asWebAuthenticationSession = new AuthenticationServices.ASWebAuthenticationSession(new NSUrl(authorizationUri.AbsoluteUri),
+                        redirectUri.Scheme, (callbackUrl, error) =>
+                        {
+                            if (error != null)
+                            {
+                                ProcessCompletionHandlerError(error);
+                            }
+                            else
+                            {
+                                vc.InvokeOnMainThread(() =>
+                                {
+                                    ContinueAuthentication(callbackUrl.ToString());
+                                });
+                            }
+                        });
+
+                    asWebAuthenticationSession.Start();
+                }
+
+                else if (UIDevice.CurrentDevice.CheckSystemVersion(11, 0))
                 {
-                    vc.PresentViewController(safariViewController, false, null);
-                });
+                    sfAuthenticationSession = new SFAuthenticationSession(new NSUrl(authorizationUri.AbsoluteUri),
+                        redirectUri.Scheme, (callbackUrl, error) =>
+                        {
+                            if (error != null)
+                            {
+                                ProcessCompletionHandlerError(error);
+                            }
+                            else
+                            {
+                                vc.InvokeOnMainThread(() =>
+                                {
+                                    ContinueAuthentication(callbackUrl.ToString());
+                                });
+                            }
+                        });
+
+                    sfAuthenticationSession.Start();
+                }
+
+                else
+                {
+                    safariViewController = new SFSafariViewController(new NSUrl(authorizationUri.AbsoluteUri), false);
+                    safariViewController.Delegate = this;
+                    vc.InvokeOnMainThread(() =>
+                    {
+                        vc.PresentViewController(safariViewController, false, null);
+                    });
+                }
             }
             catch (Exception ex)
             {
                 requestContext.Logger.ErrorPii(ex);
                 throw CoreExceptionFactory.Instance.GetClientException(
-                    CoreErrorCodes.AuthenticationUiFailedError, 
-                    "Failed to invoke SFSafariViewController", 
+                    CoreErrorCodes.AuthenticationUiFailedError,
+                    "Failed to invoke SFSafariViewController",
                     ex);
             }
         }
 
+        public void ProcessCompletionHandlerError(NSError error)
+        {
+            if (returnedUriReady != null)
```

Comment:
learning for me:  I see that this sets the authorizationResult onto the class, so someone can look at it.  But how does this trigger an error/exception callback to the caller that their async operation has failed?

also, what is returnedUriReady, how/when can it be null, and does the user not get notified of a failure if it is?

Reply:
the authResult gets pushed back through the code to the InteractiveRequest, where it is verified (VerifiyAuthorizationResult) and if the result has AuthorizationStatus.UserCancel, we throw an MsalClientException there.

returnedUriReady is a semaphoreSlim, and if it's null, it goes back to the app in the AppDelegate, OpenUrl.

Reply:
Awesome.  It's not clear just from reading this method how the error propagates.  Perhaps adding a comment here explaining where this propagates to the exception throw would be helpful.